[Unified PDF] [iOS] Safari loading blank PDF document when served with Content-Disposition: attachment HTTP header

aprotyas opened the pull request:
#### 9978479e764ea5982422797cfc2ea513c098269e
<pre>
[Unified PDF] [iOS] Safari loading blank PDF document when served with Content-Disposition: attachment HTTP header
<a href="https://bugs.webkit.org/show_bug.cgi?id=288432">https://bugs.webkit.org/show_bug.cgi?id=288432</a>
<a href="https://rdar.apple.com/143879793">rdar://143879793</a>

Reviewed by Tim Horton.

If a PDF file is served with the `Content-Disposition: attachment` HTTP
header, then we place sandbox restrictions on said file in Cocoa ports.

In macOS, Safari provides a download prompt for the file. In iOS,
without Unified PDF enabled, the file is display inline regardless for
main frame documents, while &lt;embed&gt;/&lt;iframe&gt; elements cause a download
prompt to be presented. However, with Unified PDF enabled, we  now sink
the data in DocumentWriter as we are about to create a PluginDocument,
leaving us with blank content instead for main frame documents.

This patch bridges the gap between shipping and Unified PDF on iOS, by
freeing the PluginDocument off of plugin sandbox restrictions that we
choose to apply as part of our general strategy for Content-Disposition:
attachment headers.

Tested manually for now. <a href="https://webkit.org/b/288435">https://webkit.org/b/288435</a> tracks landing an
automated test for this.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::applyContentDispositionAttachmentSandbox):

Canonical link: <a href="https://commits.webkit.org/291285@main">https://commits.webkit.org/291285@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/985abbdff1818f57acf63e6dada832e23d661daa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92552 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1706 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/43059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94602 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20555 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/43059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95554 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99563 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79674 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/79179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14742 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21014 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9710,10 +9710,13 @@ void Document::applyContentDispositionAttachmentSandbox()
+
+    if (isMediaDocument())
         enforceSandboxFlags(SandboxFlag::Origin);
+    else if (isPluginDocument())
```
What kind of documents (or what kind of content) will return true for `isPluginDocument()`? You are changing the sandbox for all of them. Even if plugin documents are limited to PDFs today, are they guaranteed to be in the future?

hortont424 replied:
I doubt we will be reexpanding support for plugins past PDF (and in fact this assumption has already spread in the code; see e.g. WebKit2's PluginView knowing that it is only a PDF).

aprotyas replied:
It's highly unlikely that WebKit2's PluginView becomes non-exclusive to PDF content. At that point, we would have to revisit a lot of assumptions baked into our code.

johnwilander commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9710,10 +9710,13 @@ void Document::applyContentDispositionAttachmentSandbox()
+    if (isMediaDocument())
         enforceSandboxFlags(SandboxFlag::Origin);
+    else if (isPluginDocument())
+        enforceSandboxFlags(SandboxFlags::all() - SandboxFlag::Plugins);
```
This looks weird in that read out loud, this code says "If it's a plugin document, remove the plugin sandbox flag." Is there a way to make this obviously correct when you read it?

hortont424 replied:
Isn't that exactly what you're doing? The outloud reading sounds exactly right.

aprotyas replied:
I was thinking of a way to alleviate the counterintuitiveness for readers, but the best I could do was an explanatory comment:

```cpp
    auto sandboxFlags = [this] -> SandboxFlags {
        if (isMediaDocument())
            return SandboxFlag::Origin;
        if (isPluginDocument()) {
            // This reads counterintuitively, but for PluginDocuments, we want all other
            // sandbox restrictions to still apply, inspite of the fact that the sandbox
            // should _not_ restrict plugin instantiation.
            SandboxFlags::all() - SandboxFlag::Plugins;
        }
        return SandboxFlags::all();
    }();
```

johnwilander reviewed (commented):
> Tested manually for now. https://webkit.org/b/288435 tracks landing an automated test for this.

What did you test? When relaxing security protections, negative testing is especially important. A positive test will pass by removing the whole sandbox which is not what we want. I'd prefer if the automated test case was included here and that it showed that the sandbox is intact except for the narrow opening you are introducing.

aprotyas committed:
[Unified PDF] [iOS] Safari loading blank PDF document when served with Content-Disposition: attachment HTTP header
https://bugs.webkit.org/show_bug.cgi?id=288432
rdar://143879793

Reviewed by Tim Horton.

If a PDF file is served with the `Content-Disposition: attachment` HTTP
header, then we place sandbox restrictions on said file in Cocoa ports.

In macOS, Safari provides a download prompt for the file. In iOS,
without Unified PDF enabled, the file is display inline regardless for
main frame documents, while <embed>/<iframe> elements cause a download
prompt to be presented. However, with Unified PDF enabled, we  now sink
the data in DocumentWriter as we are about to create a PluginDocument,
leaving us with blank content instead for main frame documents.

This patch bridges the gap between shipping and Unified PDF on iOS, by
freeing the PluginDocument off of plugin sandbox restrictions that we
choose to apply as part of our general strategy for Content-Disposition:
attachment headers.

Tested manually for now. https://webkit.org/b/288435 tracks landing an
automated test for this.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::applyContentDispositionAttachmentSandbox):

Canonical link: https://commits.webkit.org/291285@main
